### PR TITLE
Utilize TSAI tiler to crop out FOVs

### LINKED ISSUE #12: Automatic identification of RNCM position for TMA cores
**Is your feature request related to a problem? Please describe.**
The current pipeline extracts whole TIFF images for each peak. For TMAs, however, the entire TMA is extracted at once. In order to FOV-level statistics, each core needs to be identified from the single extracted tif.

**Describe the solution you'd like**
A tool to automatically identify and crop out each TMA core in a given image. These coordinates can then be reused across all of the channels from a given TMA to produce FOV-specific crops. 

One possibility would be looking into using Albert's TMA tiling tool to accomplish this. Another option would be building something ourselves to identify the RNCM postions. 

Following this step, we'd need to actually crop out each FOV. This would likely be fairly straightforward with some thresholding/image processing, as the image that gets produced has tons of contrast: 
image

### PR DESCRIPTION
**Pull request by alex-l-kong**
**What is the purpose of this PR?**

Closes #12.

**How did you implement your changes**

Using the `.imzml` file, and combined with the output from TSAI:

1. Assign each scanned coordinate on the slide to their respective core, as labeled by TSAI
2. Use the coordinates to crop out the FOVs

**Remaining issues**

Currently, `generate_glycan_mask` segments out all of the cores at once. The result of `map_coordinates_to_core_name`, however, can be used to segment out individual cores. We still need to update `generate_glycan_mask` to handle individual core segmentation.

### CODE REVIEW

File: conftest.py [hunk 1]
```diff
@@ -114,3 +117,82 @@ def image_xr(rng: np.random.Generator, library: pd.DataFrame) -> Generator[xr.Da
...
+    io.imsave(glycan_img_file, glycan_img)
+
+    yield glycan_img_file
+
+
+@pytest.fixture(scope="session")
+def poslog_path(tmp_path_factory: TempPathFactory, imz_data: ImzMLParser):
+    columns_write: List[str] = ["Date", "Time", "Region", "PosX", "PosY", "X", "Y", "Z"]
+    poslog_data: pd.DataFrame = pd.DataFrame(
+        np.random.rand(len(imz_data.coordinates) + 2, len(columns_write)), columns=columns_write
```
Comment (srivarra) [inline conftest.py]: I believe `np.random.default_rng` is the recommended way to generate random numbers in NumPy as of late.

We can use the `rng` fixture to generate the random data 

```python
rng.random(size=(len(imz_data.coordinates) + 2, len(columns_write))]
```

File: src/maldi_tools/extraction.py [hunk 2]
```diff
@@ -337,3 +340,90 @@ def library_matching(
...
+        pd.DataFrame:
+            Maps each coordinate to their core
+    """
+    validate_paths([centroid_path, poslog_path])
+
+    coords = np.array([coord[:2] for coord in imz_data.coordinates])
+    region_core_info = pd.read_csv(
+        poslog_path,
+        delimiter=" ",
+        names=["Date", "Time", "Region", "PosX", "PosY", "X", "Y", "Z"],
```
Comment (srivarra) [inline src/maldi_tools/extraction.py]: Do we need to read in the `"Z"` coordinate?

**Review by srivarra — CHANGES_REQUESTED** [review]
Just one small aesthetic thing, and a question about `"Z"` coordinates.

File: templates/maldi-pipeline.ipynb [hunk 3]
```diff
@@ -336,9 +336,7 @@
   {
```
Comment (srivarra) [inline templates/maldi-pipeline.ipynb:336]: <pre class="ql-syntax" spellcheck="false">Line #5.    # cores_to_crop = region_core_info["Core"].unique().tolist()
</pre><p>We can make this a bit easier for the user, maybe inside the <code>extraction.generate_glycan_mask</code> function allow <code>cores_to_crop</code> to optionally be <code>"all"</code> and we can move this complexity into the function.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/maldi-tools/pull/20/discussion/#comment-1401119716'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='56'/>

**Review by srivarra — APPROVED** [review]
Looks good to me!

**Comment by alex-l-kong** [comment]
@kxleow ran into a case where multiple slides need to be supported. That is currently not the case, as we assume the user has only one poslog file per run. This will need to be added into the pipeline.

**Comment by alex-l-kong** [comment]
Cropping now supports multiple poslog files.

**Comment by kxleow** [comment]
find nearest neighbour pixel
veri

**Comment by kxleow** [comment]
Alternative to using poslog and imzml for core cropping:
1. After generating binary mask, use imageJ to segment/split up cores into individual binary masks
2. Then use centroids in json to assign core name or positions to each mask
3. Output table - a column with mask ids and a column with corresponding core names
Please create a new branch, perhaps a separate notebook for this function. Thanks!

Sub-issue: Find nearest neighbours if centroid not found
- Encountered a case where the centroid fell exactly on a point that was not acquired in a core region. 
- Would be helpful to consider these cases by testing for a few neighbouring points if the centroid is not found in any cores
- to assist with troubleshooting, would also be useful to print which centroid the function failed to match

**Comment by kxleow** [comment]
Ran into a case with >4 TMAs combined and the order of TMAs did not follow a sequential order in the image (left-right or top-bottom). The order of TMAs were found in the warnings generated at the core cropping step.

**Comment by alex-l-kong** [comment]
> Ran into a case with >4 TMAs combined and the order of TMAs did not follow a sequential order in the image (left-right or top-bottom). The order of TMAs were found in the warnings generated at the core cropping step.

@kxleow the updated spectra extractor will prevent this issue from popping up once that gets merged in.

**Pull request merged**